File: facebook/src/com/facebook/RequestProgress.java
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook;

import android.os.Handler;

class RequestProgress {
    private final GraphRequest request;
    private final Handler callbackHandler;
    private final long threshold;

    private long progress, lastReportedProgress, maxProgress;

    RequestProgress(Handler callbackHandler, GraphRequest request) {
        this.request = request;
        this.callbackHandler = callbackHandler;

        this.threshold = FacebookSdk.getOnProgressThreshold();
    }

    long getProgress() {
        return progress;
    }

    long getMaxProgress() {
        return maxProgress;
    }

    void addProgress(long size) {
        progress += size;

        if (progress >= lastReportedProgress + threshold || progress >= maxProgress) {
            reportProgress();
        }
    }

    void addToMax(long size) {
        maxProgress += size;
    }

    void reportProgress() {
        if (progress > lastReportedProgress) {
            GraphRequest.Callback callback = request.getCallback();
            if (maxProgress > 0 && callback instanceof GraphRequest.OnProgressCallback) {
                // Keep copies to avoid threading issues
                final long currentCopy = progress;
                final long maxProgressCopy = maxProgress;
                final GraphRequest.OnProgressCallback callbackCopy =
                        (GraphRequest.OnProgressCallback) callback;
                if (callbackHandler == null) {
                    callbackCopy.onProgress(currentCopy, maxProgressCopy);
                }
                else {
                    callbackHandler.post(new Runnable() {
                        @Override
                        public void run() {
                            callbackCopy.onProgress(currentCopy, maxProgressCopy);
                        }
                    });
                }
                lastReportedProgress = progress;
            }
        }
    }
}

File: facebook-core/src/main/java/com/facebook/RequestProgress.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook

import android.os.Handler

class RequestProgress
constructor(private val callbackHandler: Handler?, private val request: GraphRequest) {
  private val threshold = FacebookSdk.getOnProgressThreshold()
  var progress: Long = 0
    private set
  private var lastReportedProgress: Long = 0
  var maxProgress: Long = 0
    private set

  /**
   * Increase process size
   *
   * @param size to increase
   */
  fun addProgress(size: Long) {
    progress += size
    if (progress >= lastReportedProgress + threshold || progress >= maxProgress) {
      reportProgress()
    }
  }

  /**
   * Increase maxProgress size
   *
   * @param size to increase
   */
  fun addToMax(size: Long) {
    maxProgress += size
  }

  /** Send requests if process size larger than lastReportedProgress size */
  fun reportProgress() {
    if (progress > lastReportedProgress) {
      val callback = request.callback
      if (maxProgress > 0 && callback is GraphRequest.OnProgressCallback) {
        // Keep copies to avoid threading issues
        val progressCopy = progress
        val maxProgressCopy = maxProgress
        callbackHandler?.post { callback.onProgress(progressCopy, maxProgressCopy) }
            ?: callback.onProgress(progressCopy, maxProgressCopy)
        lastReportedProgress = progress
      }
    }
  }
}
